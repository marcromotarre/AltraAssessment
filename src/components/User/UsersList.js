import React, { useState } from 'react';
import { Link } from '@reach/router';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import UserElement from './UserElement.js';
import UserPagination from './UserPagination.js';

export const UsersListContainer = styled.div`
  a {
    text-decoration: none;
  }
`;

const UsersList = ({ users, usersPerPage }) => {
  const [page, setPage] = useState(1);

  const loadMoreUsers = () => {
    setPage(page + 1);
  };

  const moreUsers = usersPerPage * page >= users.length ? false : true;

  return (
    <UsersListContainer>
      {users &&
        users
          .filter((a, index) => index < usersPerPage * page)
          .map((user) => (
            <Link key={user.id} to={`/user/${user.id}`}>
              <UserElement {...user} />
            </Link>
          ))}
      {moreUsers && <UserPagination onClick={loadMoreUsers} />}
    </UsersListContainer>
  );
};

UsersList.propTypes = {
  users: PropTypes.array,
  usersPerPage: PropTypes.number,
};

UsersList.defaultProps = {
  users: [],
  usersPerPage: Number.POSITIVE_INFINITY,
};

export default UsersList;
