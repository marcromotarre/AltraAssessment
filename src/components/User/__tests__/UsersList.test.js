import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import UsersList from '../UsersList';
import { shallow } from 'enzyme';
import { fireEvent } from '@testing-library/react';

import users from './../../../fixtures/users';

/** this component should:
 *  - show all users information using UserElement component
 *      - all users has Link to user site by id
 *
 *  - at the end of the list paint user Pagination button if there are more users to paint
 */

const loadMoreUsers = jest.fn();

describe('Users List Component tests', () => {
  it('render UsersList Snapshot', () => {
    const tree = renderer
      .create(<UsersList users={users} usersPerPage={2} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('UsersList should paint all users', () => {
    const wrapper = shallow(<UsersList users={users} />);
    expect(wrapper.children().length).toBe(users.length);
  });

  it('UsersList should paint Link and UserElement for each user', () => {
    const wrapper = shallow(<UsersList users={users} />);
    const index = 3;
    const iUser = wrapper.childAt(index);
    expect(iUser.children().length).toBe(1);
    expect(iUser.find('Link').prop('to')).toBe(`/user/${users[index].id}`);
    expect(iUser.find('Link').children().length).toBe(1);
    const userElement = iUser.find('Link').childAt(0);
    expect(userElement.prop('id')).toBe(users[index].id);
    expect(userElement.prop('name')).toBe(users[index].name);
    expect(userElement.prop('thumbnail')).toBe(users[index].thumbnail);
    expect(userElement.prop('age')).toBe(users[index].age);
    expect(userElement.prop('weight')).toBe(users[index].weight);
    expect(userElement.prop('height')).toBe(users[index].height);
    expect(userElement.prop('hair_color')).toBe(users[index].hair_color);
    expect(userElement.prop('professions')).toBe(users[index].professions);
    expect(userElement.prop('friends')).toBe(users[index].friends);
  });

  it('Users pagination should be painted if there are more users than users per page', () => {
    const wrapper = shallow(
      <UsersList users={users} usersPerPage={users.length - 1} />
    );
    const userPagination = wrapper.childAt(users.length - 1);
    expect(wrapper.children().length).toBe(users.length);
    expect(userPagination.prop('text')).toBe('Load more users');
  });

  it('Users pagination should not be painted if there are same users than users per page', () => {
    const wrapper = shallow(
      <UsersList users={users} usersPerPage={users.length} />
    );
    expect(wrapper.children().length).toBe(users.length);
  });

  it('Users pagination should not be painted if there are less users than users per page', () => {
    const wrapper = shallow(
      <UsersList users={users} usersPerPage={users.length + 1} />
    );
    expect(wrapper.children().length).toBe(users.length);
  });

  it('click on User Pagination', () => {
    const { container } = render(<UsersList users={users} usersPerPage={2} />);
    expect(container.querySelectorAll('a').length).toBe(2);
    expect(container.querySelectorAll('button').length).toBe(1);
    const paginationButton = container.querySelector('button');
    fireEvent.click(paginationButton);
    expect(container.querySelectorAll('a').length).toBe(4);
    fireEvent.click(paginationButton);
    expect(container.querySelectorAll('a').length).toBe(5);
    expect(container.querySelectorAll('button').length).toBe(0);
  });
});
