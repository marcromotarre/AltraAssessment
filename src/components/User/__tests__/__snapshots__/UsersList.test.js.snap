// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users List Component tests render UsersList Snapshot 1`] = `
<div
  className="sc-AxmLO gmopCH"
>
  <a
    href="/user/0"
    onClick={[Function]}
  >
    <section
      className="sc-AxiKw gZriPm"
    >
      <img
        className="sc-AxhCb eTSSEI"
        data-testid="user-image"
        src="image1"
      />
      <div
        className="sc-AxhUy QOkiE"
      >
        <label
          className="sc-AxgMl ijardh"
          data-testid="user-name"
        >
          one
        </label>
        <div
          className="sc-AxheI jYVyby"
          data-testid="user-properties"
        >
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              width={14.63}
            />
             
            <label>
              10
            </label>
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              height={17.7}
            />
            <label>
              50
            </label>
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              width={17.7}
            />
            <label>
              40
            </label>
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              className="sc-AxirZ eMVPxk"
              color="Red"
              width={15.26}
            />
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              width={14.34}
            />
            <label>
              3
            </label>
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              width={15.26}
            />
             
            <label>
              2
            </label>
          </div>
        </div>
      </div>
    </section>
  </a>
  <a
    href="/user/1"
    onClick={[Function]}
  >
    <section
      className="sc-AxiKw gZriPm"
    >
      <img
        className="sc-AxhCb eTSSEI"
        data-testid="user-image"
        src="image2"
      />
      <div
        className="sc-AxhUy QOkiE"
      >
        <label
          className="sc-AxgMl ijardh"
          data-testid="user-name"
        >
          two
        </label>
        <div
          className="sc-AxheI jYVyby"
          data-testid="user-properties"
        >
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              width={14.63}
            />
             
            <label>
              15
            </label>
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              height={17.7}
            />
            <label>
              70
            </label>
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              width={17.7}
            />
            <label>
              40
            </label>
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              className="sc-AxirZ itfaLk"
              color="Green"
              width={15.26}
            />
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              width={14.34}
            />
            <label>
              1
            </label>
          </div>
          <div
            className="sc-AxjAm cDIZLj"
            data-testid="user-property"
          >
            <svg
              width={15.26}
            />
             
            <label>
              1
            </label>
          </div>
        </div>
      </div>
    </section>
  </a>
  <button
    className="sc-Axmtr bXIkFr"
    onClick={[Function]}
  >
    <label>
      Load more users
    </label>
  </button>
</div>
`;
